src/validators/arguments_v3.rs

```diff
@@ -0,0 +1,654 @@
+use std::str::FromStr;
+
+use pyo3::intern;
+use pyo3::prelude::*;
+use pyo3::types::{PyDict, PyList, PyString, PyTuple};
+
+use ahash::AHashSet;
+use pyo3::IntoPyObjectExt;
+
+use crate::build_tools::py_schema_err;
+use crate::build_tools::{schema_or_config_same, ExtraBehavior};
+use crate::errors::{ErrorTypeDefaults, ValError, ValLineError, ValResult};
+use crate::input::{
+    Arguments, BorrowInput, Input, KeywordArgs, PositionalArgs, ValidatedDict, ValidatedTuple, ValidationMatch,
+};
+use crate::lookup_key::LookupKeyCollection;
+use crate::tools::SchemaDict;
+
+use super::validation_state::ValidationState;
+use super::{build_validator, BuildValidator, CombinedValidator, DefinitionsBuilder, Validator};
+
+#[derive(Debug, PartialEq)]
+enum ParameterMode {
+    PositionalOnly,
+    PositionalOrKeyword,
+    VarArgs,
+    KeywordOnly,
+    VarKwargsUniform,
+    VarKwargsUnpackedTypedDict,
+}
+
+impl FromStr for ParameterMode {
+    type Err = PyErr;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "positional_only" => Ok(Self::PositionalOnly),
+            "positional_or_keyword" => Ok(Self::PositionalOrKeyword),
+            "var_args" => Ok(Self::VarArgs),
+            "keyword_only" => Ok(Self::KeywordOnly),
+            "var_kwargs_uniform" => Ok(Self::VarKwargsUniform),
+            "var_kwargs_unpacked_typed_dict" => Ok(Self::VarKwargsUnpackedTypedDict),
+            s => py_schema_err!("Invalid var_kwargs mode: `{}`", s),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct Parameter {
+    name: String,
+    mode: ParameterMode,
+    lookup_key_collection: LookupKeyCollection,
+    validator: CombinedValidator,
+}
+
+impl Parameter {
+    fn is_variadic(&self) -> bool {
+        matches!(
+            self.mode,
+            ParameterMode::VarArgs | ParameterMode::VarKwargsUniform | ParameterMode::VarKwargsUnpackedTypedDict
+        )
+    }
+}
+
+#[derive(Debug)]
+pub struct ArgumentsV3Validator {
+    parameters: Vec<Parameter>,
+    positional_params_count: usize,
+    loc_by_alias: bool,
+    extra: ExtraBehavior,
+    validate_by_alias: Option<bool>,
+    validate_by_name: Option<bool>,
```

Comment:
do we really need both these options at schema build time?

I can see users wanting to use `by_alias` and `by_name` at validation time, but setting these now seems unlikely?

Reply:
I matched it with the current arguments schema, which was updated by Sydney's recent work on aliases configuration and so it was just easier to mirror it.

Anyway, I think 95% of users are using `@validate_call` and aren't using a `TypeAdapter`/the core schemas directly. It's probably easier to leave it as is as the implementation and tests are already there?

Reply:
I think there is no harm to leave it as is.